tests/mappings/mapping_q_eulerian_09.cc

```diff
@@ -0,0 +1,187 @@
+  DoFTools::make_hanging_node_constraints  (dof_handler, constraints);
+  constraints_euler.close ();
+
+  // Displacement vector
+  LinearAlgebra::distributed::Vector<NumberType> displacement;
+  displacement.reinit(locally_owned_dofs_euler,
+                      locally_relevant_dofs_euler,
+                      mpi_communicator);
+  displacement = 0.;
+
+  // Find the vertex at the origin and Euler DoFs at this vertex
+  const Point<dim> target;
+  typename DoFHandler<dim>::active_cell_iterator
+  euler_cell = GridTools::find_active_cell_around_point (dof_handler_euler,target);
+  unsigned int best_vertex   = 0;
+  double       best_dist     = target.distance(euler_cell->vertex(best_vertex));
+
+  for (unsigned int vertex_no = 1; vertex_no < GeometryInfo<dim>::vertices_per_cell; vertex_no++)
+    {
+      const double dist = target.distance(euler_cell->vertex(vertex_no));
+      if (dist < best_dist)
+        {
+          best_vertex = vertex_no;
+          best_dist   = dist;
+        }
+    }
+
+  const unsigned int x_dof = euler_cell->vertex_dof_index(best_vertex, 0);
+
+  const double x_displacement = 0.1;
+
+  // first, move via mapping:
+  displacement(x_dof) = x_displacement;
+  displacement.compress(VectorOperation::insert);
+
+  // The core : compute matrix-vector products with mass and laplace operators
+  QGauss<1> quadrature_formula(n_q_points);
+
+  MappingQEulerian<dim,LinearAlgebra::distributed::Vector<NumberType>>
+      euler_mapping(euler_fe_degree, dof_handler_euler, displacement);
+
+  std::shared_ptr<MatrixFree<dim,NumberType>> matrix_free(new MatrixFree<dim,NumberType> ());
+  typename MatrixFree<dim,NumberType>::AdditionalData data;
+  data.tasks_parallel_scheme = MatrixFree<dim,NumberType>::AdditionalData::partition_color;
+  data.mapping_update_flags = update_values | update_gradients | update_JxW_values;
+  matrix_free->reinit (euler_mapping, dof_handler, constraints, quadrature_formula, data);
+
+  LinearAlgebra::distributed::Vector<NumberType> src, dst1, dst2, dst3, dst4;
+  matrix_free->initialize_dof_vector(src);
+  matrix_free->initialize_dof_vector(dst1);
+  matrix_free->initialize_dof_vector(dst2);
+  matrix_free->initialize_dof_vector(dst3);
+  matrix_free->initialize_dof_vector(dst4);
+
+  for (unsigned int i = 0; i < src.local_size(); ++i)
+    src.local_element(i) = (NumberType)Testing::rand()/RAND_MAX;
+
+  MatrixFreeOperators::MassOperator<dim,fe_degree, n_q_points, 1, LinearAlgebra::distributed::Vector<NumberType>> mf_mass;
+  MatrixFreeOperators::MassOperator<dim,fe_degree, n_q_points, 1, LinearAlgebra::distributed::Vector<NumberType>> mf_laplace;
+  mf_mass.initialize(matrix_free);
+  mf_laplace.initialize(matrix_free);
+
+  mf_mass.vmult(dst1,src);
+  mf_laplace.vmult(dst3,src);
+
+  // now move manually
+  displacement = 0.;
+  Point<dim> &v = euler_cell->vertex(best_vertex);
+  v(0) += x_displacement;
+  matrix_free->reinit (euler_mapping, dof_handler, constraints, quadrature_formula, data);
```

Comment:
@kronbichler I presume this resets whatever was cached within the `MatrixFree` object with the previous combination of Mapping (u!=0) + Triangulation (unmodified).

Reply:
@davydden Yes, this does re-compute the Jacobians etc. that get cached in `MatrixFree`. To minimize the data that is re-computed, one could also set the `MatrixFree::AdditionalData::initialize_indices` to false to keep the previous content, see here:
https://github.com/dealii/dealii/blob/master/tests/matrix_free/update_mapping_only.cc

Reply:
thanks for confirming!